### PR TITLE
addressing sync over async issues

### PR DESCRIPTION
**Pull request by mhowlett**
@edenhill - ready for review.

notes:
- removes all sync-over-async functionality from the client class implementations.
- provides adapter classes to enable use of async serdes in sync scenerios.
  - previously in the client classes, Task.Run was used to ensure there were no single threaded synchronization context issues.
  - now the sync-over-async is explicit, it's more appropriate to just use .GetResult() (not requiring additional thread pool usage for safety).
- revert RC4 (sync serde interface change) to enable the above.
  - note: extension methods could be used to accommodate SimpleSerializer/SimpleDeserializer delegates in the future without making changes to the builder/client classes.
- the only no-compromise solution for async deserializers is to implement `ConsumeAsync` (and friends).
  - before then, the only way to use async deserializers in the consumer is via the sync adapter.
  - we could implement purely sync SR integration, but it would make use of older library functions and at this point would look fairly antiquated.
- added some missed `.ConfigureAwait(false)`'s in SR client that were errantly omitted previously.
- Includes a console app I was using to investigate deadlock scenarios (easier than inside the test framework). I could exclude this, but I think it's useful to keep it (though marginal).

### CODE REVIEW

File: test/Confluent.SchemaRegistry.Serdes.IntegrationTests/Tests/SyncOverAsync.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
+// Copyright 2016-2017 Confluent Inc.
```
Comment (edenhill) [inline test/Confluent.SchemaRegistry.Serdes.IntegrationTests/Tests/SyncOverAsync.cs]: 2019

File: examples/Protobuf/Program.cs [hunk 2]
```diff
@@ -29,6 +29,31 @@
 /// </summary>
 namespace Confluent.Kafka.Examples.Protobuf
 {
+    /// <summary>
+    ///     protobuf serializer
+    /// </summary>
+    public class ProtobufSerializer<T> : ISerializer<T> where T : IMessage<T>, new()
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply examples/Protobuf/Program.cs]: The convention is only async is explicitly specified. If you don't see Async, you assume it's sync. sync is usually implicit both in method names (common, suffix) and class names (less common, prefix, for example AsyncEnumerator vs Enumerator).

**Comment by mhowlett** [comment]
I'm thinking just `Produce` is the best name for `BeginProduce`, but I'll sleep on it.

File: src/Confluent.Kafka/SyncOverAsyncDeserializer.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace Confluent.Kafka
+{
+    /// <summary>
+    ///     An adapter that allows an async deserializer
+    ///     to be used where a sync deserializer is required.
+    ///     In using this adapter, there are two potential
+    ///     issues you should be aware of:
+    /// 
+    ///     1. If you are working in a single threaded
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply src/Confluent.Kafka/SyncOverAsyncDeserializer.cs]: single threaded is important - if the sync context isn't single threaded, there are generally many other threads available for tasks to complete on, and generally there will be no deadlock (except in the case of thread pool starvation).

File: examples/Protobuf/Program.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhowlett) [reply examples/Protobuf/Program.cs]: same comment in response to your other sync comments.

[the remark this answers is not in the document]
Comment (mhowlett) [reply examples/Protobuf/Program.cs]: btw: we can make this all nice and terse again in 1.0.1 by having extension methods that allow these to be specified as delegates again.

**Pull request merged**